### PR TITLE
MAINT: Remove duplicate implementation for aliased functions.

### LINKED ISSUE #10651: BUG: np.prod and np.product behave differently on duck-typed arrays
```python
>>> class CanProd(object):
...	def prod(self, axis, dtype, out): return "prod"

>>> np.product(CanProd())
<__main__.CanProd object at 0x0000023BAF7B29E8>
>>> np.prod(CanProd())
'prod'
```

We should probably deprecate one of these, or simply make them identical.

The docs already claim they are the same

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #10651 by replacing the behaviour of `product` with that of `prod`.

Also group all the aliases together.

### CODE REVIEW

File: numpy/core/fromnumeric.py [hunk 1]
```diff
@@ -3232,3 +3115,105 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue):
...
+    """
+    Return the number of dimensions of an array.
+
+    If `a` is not already an array, a conversion is attempted.
+    Scalars are zero dimensional.
+
+    .. note::
+        This function is deprecated in NumPy 1.9 to avoid confusion with
+        `numpy.linalg.matrix_rank`. The ``ndim`` attribute or function
+        should be used instead.
```
Comment (eric-wieser) [inline numpy/core/fromnumeric.py]: I wonder if we should just drop all of this docstring after this line

**Comment by eric-wieser** [comment]
"There should be one-- and preferably only one --obvious way to do it." suggests we should deprecate these aliases, but that's out of scope for now

File: numpy/core/fromnumeric.py (hunk 1 above)
Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: I'm +1 on that. Just maybe put an equivalent to in see also?

Comment (hameerabbasi) [reply numpy/core/fromnumeric.py]: Actually, it would be better to make it mirror the definition of the other aliases, but with that note.

**Review by charris — APPROVED** [review]
LGTM, go ahead and merge when you decide what you want to do with `rank`

**Comment by eric-wieser** [comment]
Alright, trimmed the `rank` docstring too - the longer it is, the more it legitimizes it and makes it appear in google results.

**Comment by charris** [comment]
Thanks Eric.

**Pull request merged**